streams.go

```diff
@@ -0,0 +1,296 @@
+	chunkIndex := uint64(0)
+	for i := 0; i < len(data); i += STREAM_CHUNK_SIZE {
+		end := i + STREAM_CHUNK_SIZE
+		if end > len(data) {
+			end = len(data)
+		}
+
+		e.waitForBufferStatusLow(protocol.DataPacket_RELIABLE)
+		e.publishStreamChunk(&protocol.DataStream_Chunk{
+			StreamId:   streamId,
+			Content:    data[i:end],
+			ChunkIndex: chunkIndex,
+		}, destinationIdentities)
+
+		if onProgress != nil && totalSize != nil {
+			progress := float64(i) / float64(*totalSize)
+			(*onProgress)(progress)
+		}
+
+		chunkIndex++
+	}
+}
+
+type TextStreamWriter struct {
+	*baseStreamWriter[string]
+	Info TextStreamInfo
+}
+
+func newTextStreamWriter(info TextStreamInfo, header *protocol.DataStream_Header, e *RTCEngine, destinationIdentities []string, onProgress *func(progress float64)) *TextStreamWriter {
+	e.publishStreamHeader(header, destinationIdentities)
+
+	onWrite := func(data string) {
+		writeStreamBytes([]byte(data), e, info.Id, destinationIdentities, info.Size, onProgress)
+	}
+
+	onClose := func() {
+		e.publishStreamTrailer(info.Id, destinationIdentities)
+	}
+
+	return &TextStreamWriter{
+		baseStreamWriter: newBaseStreamWriter[string](onWrite, onClose),
+		Info:             info,
+	}
+}
+
+type ByteStreamWriter struct {
+	*baseStreamWriter[[]byte]
+	Info ByteStreamInfo
+}
+
+func newByteStreamWriter(info ByteStreamInfo, header *protocol.DataStream_Header, e *RTCEngine, destinationIdentities []string, onProgress *func(progress float64)) *ByteStreamWriter {
+	e.publishStreamHeader(header, destinationIdentities)
+
+	onWrite := func(data []byte) {
+		writeStreamBytes(data, e, info.Id, destinationIdentities, info.Size, onProgress)
+	}
+
+	onClose := func() {
+		e.publishStreamTrailer(info.Id, destinationIdentities)
+	}
+
+	return &ByteStreamWriter{
+		baseStreamWriter: newBaseStreamWriter[[]byte](onWrite, onClose),
+		Info:             info,
+	}
+}
+
+type baseStreamReader struct {
+	reader        chan *protocol.DataStream_Chunk
+	totalByteSize *uint64
+	bytesReceived int
+
+	closed     bool
+	readerLock sync.Mutex
+
+	onProgress *func(progress float64)
+}
+
+func newBaseStreamReader(totalByteSize *uint64) *baseStreamReader {
+	baseReader := &baseStreamReader{
+		reader:        make(chan *protocol.DataStream_Chunk),
+		bytesReceived: 0,
+	}
+	if totalByteSize != nil {
+		baseReader.totalByteSize = totalByteSize
+	}
+	return baseReader
+}
+
+func (r *baseStreamReader) enqueue(chunk *protocol.DataStream_Chunk) {
+	r.readerLock.Lock()
+	defer r.readerLock.Unlock()
+
+	if r.closed {
+		return
+	}
+	r.reader <- chunk
+}
+
+func (r *baseStreamReader) OnProgress(onProgress *func(progress float64)) {
+	r.onProgress = onProgress
+}
+
+func (r *baseStreamReader) close() {
+	r.readerLock.Lock()
+	defer r.readerLock.Unlock()
+
+	if !r.closed {
+		close(r.reader)
+		r.closed = true
+	}
+}
+
+type TextStreamReader struct {
+	*baseStreamReader
+	Info TextStreamInfo
+
+	// Q: what are we using this for?
```

Comment:
cc: @lukasIO for details on how this is used.

Reply:
this was used until a week back to construct the string from chunks that have been received so far, but we got rid of this, so the receivedChunks aren't needed right now. 

We might need them (or something along the lines) once we get to chunk NACKs